pkg/cloudprovider/aws/instance.go

```diff
@@ -105,6 +128,37 @@ func (p *InstanceProvider) Terminate(ctx context.Context, node *v1.Node) error {
 	return nil
 }
 
+// The intention is to remove Offerings from the provided possible InstanceTypes based on recently observed ICEs, which will
+// indirectly impact the packer to keep it from spinning its wheels on packings that are doomed to fail due to EC2 capacity constraints.
+func (p *InstanceProvider) DiscardICEdInstanceTypes(ctx context.Context, instanceTypes []InstanceType) []cloudprovider.InstanceType {
+	var cleanedInstanceTypes = []cloudprovider.InstanceType{}
+	for index, instanceType := range instanceTypes {
+		odIceZones, hasODIces := p.iceCache.Get(createIceCacheKey(instanceType.Name(), v1alpha1.CapacityTypeOnDemand))
+		spotIceZones, hasSpotIces := p.iceCache.Get(createIceCacheKey(instanceType.Name(), v1alpha1.CapacityTypeSpot))
+		if !hasODIces && !hasSpotIces {
+			cleanedInstanceTypes = append(cleanedInstanceTypes, &instanceTypes[index])
+			continue
+		}
+		cleanedOfferings := []cloudprovider.Offering{}
+		for _, offering := range instanceType.Offerings() {
+			if hasODIces && offering.CapacityType == v1alpha1.CapacityTypeOnDemand {
```

Comment:
if this short circuits (hasODIces returns false), won't this go to the `else If` and add an OD offering?

Reply:
You are right, I need to reverse the order here because the else if should only happen if we know it's not od. However, you helped me realize another important issue. If there's no spice at all for this instance type and yet this is a spot offering I need to add the offering to the cleaned slice (yet I don't if the cache has an od ice for this instance type). In trying to be too clever with the conditionals I've created this problem.